packages/frontend/core/src/modules/app-tabs-header/views/app-tabs-header.tsx

```diff
@@ -32,89 +39,143 @@ const WorkbenchTab = ({
   workbench,
   active: tabActive,
   tabsLength,
+  dnd,
+  onDrop,
 }: {
   workbench: TabStatus;
   active: boolean;
   tabsLength: number;
+  dnd?: boolean;
+  onDrop?: (data: DropTargetDropEvent<AffineDNDData>) => void;
 }) => {
   const tabsHeaderService = useService(AppTabsHeaderService);
   const activeViewIndex = workbench.activeViewIndex ?? 0;
   const onContextMenu = useAsyncCallback(
     async (viewIdx: number) => {
-      await tabsHeaderService.showContextMenu(workbench.id, viewIdx);
+      await tabsHeaderService.showContextMenu?.(workbench.id, viewIdx);
     },
     [tabsHeaderService, workbench.id]
   );
   const onActivateView = useAsyncCallback(
     async (viewIdx: number) => {
-      await tabsHeaderService.activateView(workbench.id, viewIdx);
+      await tabsHeaderService.activateView?.(workbench.id, viewIdx);
     },
     [tabsHeaderService, workbench.id]
   );
   const onCloseTab: MouseEventHandler = useAsyncCallback(
     async e => {
       e.stopPropagation();
 
-      await tabsHeaderService.closeTab(workbench.id);
+      await tabsHeaderService.closeTab?.(workbench.id);
     },
     [tabsHeaderService, workbench.id]
   );
 
+  const { dropTargetRef, closestEdge } = useDropTarget<AffineDNDData>(
+    () => ({
+      closestEdge: {
+        allowedEdges: ['left', 'right'],
+      },
+      onDrop,
+      dropEffect: 'move',
+      canDrop: ctx => {
+        if (
+          ctx.source.data.from?.at === 'app-header:tabs' &&
+          ctx.source.data.from.tabId !== workbench.id
+        ) {
+          return true;
+        }
+
+        if (ctx.source.data.entity?.type === 'doc') {
```

Comment:
Only doc is supported yet? What about collection & tag?

Reply:
will do later